### PR TITLE
Uninstall one-by-one with waiting for the status from the server not to be in "create / update / delete operation is already in progress (10102)"

### PR DESCRIPTION
**Pull request by kobymeir**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [CIAC-7603](https://jira-hq.paloaltonetworks.local/browse/CIAC-7603)

## Description
Uninstall one-by-one with waiting for the status from the server not to be in "create / update / delete operation is already in progress (10102)"

Nightly test: https://code.pan.run/xsoar/content/-/pipelines/5907768

### CODE REVIEW

File: .gitlab/ci/.gitlab-ci.bucket-upload.yml [hunk 1]
```diff
@@ -221,18 +221,18 @@ install-packs-in-server-master:
     - CLOUD_SERVERS_PATH=$(cat $CLOUD_SERVERS_FILE)
     - echo ${CLOUD_API_KEYS} > "cloud_api_keys.json"
 
+    - section_start "Clean Machine"
```
Comment (kobymeir) [inline .gitlab/ci/.gitlab-ci.bucket-upload.yml:224]: moved Anar's e2e code to run after we clean the machine

File: .gitlab/ci/.gitlab-ci.bucket-upload.yml [hunk 2]
```diff
@@ -221,18 +221,18 @@ install-packs-in-server-master:
...
+    - ./Tests/scripts/uninstall_packs_and_reset_bucket_cloud.sh || EXIT_CODE=$?
```
Comment (kobymeir) [inline .gitlab/ci/.gitlab-ci.bucket-upload.yml:225]: don't fail on errors and proceed regardless, but still remember to fail the build.

File: .gitignore [hunk 3]
```diff
@@ -93,6 +93,7 @@ test_runner.sh
 # log files
 demisto_sdk_debug.log
 demisto_sdk_debug.log.*
+*.log
```
Comment (kobymeir) [inline .gitignore:96]: excluding log files, created by the build script

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 4]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+
+    except Exception as e:
+        logging.exception(f'The request to uninstall packs has failed. Additional info: {str(e)}')
+    return False, None
+
+
+def wait_until_not_updating(client: demisto_client,
+                            attempts_count: int = 2,
+                            sleep_interval: int = 30,
+                            maximum_time_to_wait=600,
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:131]:
```suggestion
                            maximum_time_to_wait: int = 600,
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 5]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+                            ) -> bool:
     """
 
     Args:
         client (demisto_client): The client to connect to.
-        pack_id: packs id to uninstall
+        attempts_count (int): The number of attempts to install the packs.
+        sleep_interval (int): The sleep interval, in seconds, between install attempts.
+        maximum_time_to_wait(int): The maximum time to wait for the server not to be in updating mode, in seconds.
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:139]:
```suggestion
        maximum_time_to_wait (int): The maximum time to wait for the server to exit the updating mode, in seconds.
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 6]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+        if success:
+            if not updating_status:
+                return True
+            logging.debug(f"Server is still installation/updating status, sleeping for {sleep_interval} seconds.")
+        else:
+            logging.debug(f"failed to get installation/updating status, sleeping for {sleep_interval} seconds.")
+            attempts_count -= 1
+
+        # There are more attempts available, sleep and retry.
+        if attempts_count:
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:156]: Add `else`, with log that the attempts have exhausted.

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 7]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+            try:
+                logging.info(f"Uninstalling packs {pack_id}, Attempt: {attempts_count - attempt}/{attempts_count}")
+                response, status_code, headers = demisto_client.generic_request_func(client,
+                                                                                     path=f'/contentpacks/installed/{pack_id}',
+                                                                                     method='DELETE',
+                                                                                     accept='application/json',
+                                                                                     _request_timeout=None)
+
+                if 200 <= status_code < 300 and status_code != 204:
+                    logging.success(f'Pack:{pack_id} was successfully uninstalled from the server')
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:188]:
```suggestion
                    logging.success(f'Pack: {pack_id} was successfully uninstalled from the server')
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 8]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+                    break
+
+                if not attempt:
+                    raise Exception(f"Got bad status code: {status_code}, headers: {headers}")
+
+                logging.warning(f"Got bad status code: {status_code} from the server, headers:{headers}")
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:194]:
```suggestion
                logging.warning(f"Got bad status code: {status_code} from the server, headers: {headers}")
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 9]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+
+            except ApiException as ex:
+
+                if ALREADY_IN_PROGRESS in ex.body:
+                    wait_succeeded = wait_until_not_updating(client)
+                    if not wait_succeeded:
+                        raise Exception(
+                            "Failed to wait for the server to not be in installation/updating status"
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:202]:
```suggestion
                            "Failed to wait for the server to exit installation/updating status"
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 10]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+                        ) from ex
+
+                if not attempt:  # exhausted all attempts, understand what happened and exit.
+                    # Unknown exception reason, re-raise.
+                    raise Exception(f"Got {ex.status} from server, message:{ex.body}, headers:{ex.headers}") from ex
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:207]:
```suggestion
                    raise Exception(f"Got {ex.status} from server, message: {ex.body}, headers: {ex.headers}") from ex
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 11]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+            except (HTTPError, HTTPWarning) as http_ex:
+                if not attempt:
+                    raise Exception("Failed to perform http request to the server") from http_ex
+
+            # There are more attempts available, sleep and retry.
+            logging.debug(f"failed to uninstall packs: {pack_id}, sleeping for {sleep_interval} seconds.")
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:213]:
```suggestion
            logging.debug(f"failed to uninstall pack: {pack_id}, sleeping for {sleep_interval} seconds.")
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 12]
```diff
@@ -244,7 +364,7 @@ def options_handler():
 def main():
     install_logging('cleanup_cloud_instance.log', logger=logging)
 
-    # in cloud we dont use demisto username
+    # In Cloud, We don't use demisto username
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:367]:
```suggestion
    # In Cloud, we don't use demisto username
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 13]
```diff
@@ -259,6 +379,9 @@ def main():
                                       verify_ssl=False,
                                       api_key=api_key,
                                       auth_id=xdr_auth_id)
+    client.api_client.user_agent = generate_user_agent(options.build_number)
+    logging.debug(f'Setting user agent on client to:{client.api_client.user_agent}')
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:383]:
```suggestion
    logging.debug(f'Setting user agent on client to: {client.api_client.user_agent}')
```

File: Tests/configure_and_test_integration_instances.py [hunk 14]
```diff
@@ -85,6 +84,10 @@ class Running(IntEnum):
     WITH_LOCAL_SERVER = 2
 
 
+def generate_user_agent(build_number):
```
Comment (tkatzir) [inline Tests/configure_and_test_integration_instances.py:87]: Just a suggestion. What's wrong with that name?
```suggestion
def get_custom_user_agent(build_number):
```

**Review by tkatzir — COMMENTED** [review]
Looks good!
Minor comments.

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 15]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+                                                                                     _request_timeout=None)
+
+                if 200 <= status_code < 300 and status_code != 204:
+                    # the endpoint simply returns a string (rather than a json object.)
+                    updating_status = 'true' in str(response).lower()
+                    logging.info(f"Got updating status:{updating_status}")
+                    return True, updating_status
+
+                if not attempt:
+                    raise Exception(f"Got bad status code: {status_code}, headers: {headers}")
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply Tests/Marketplace/search_and_uninstall_pack.py:107]: no

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 16]
```diff
@@ -77,23 +83,136 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+            if not updating_status:
+                return True
+            logging.debug(f"Server is still installation/updating status, sleeping for {sleep_interval} seconds.")
+        else:
+            logging.debug(f"failed to get installation/updating status, sleeping for {sleep_interval} seconds.")
+            attempts_count -= 1
+
+        # There are more attempts available, sleep and retry.
+        if attempts_count:
+            sleep(sleep_interval)
```
Comment (kobymeir) [inline Tests/Marketplace/search_and_uninstall_pack.py:157]:
```suggestion
log me
            sleep(sleep_interval)
```

File: Tests/Marketplace/search_and_uninstall_pack.py (hunk 12 above)
Comment (kobymeir) [reply Tests/Marketplace/search_and_uninstall_pack.py:367]: my grammar is correct :)
https://www.grammarly.com/blog/capitalization-after-colons

File: Tests/configure_and_test_integration_instances.py [hunk 17]
```diff
@@ -127,7 +127,7 @@ def reconnect_client(self):
                                                  verify_ssl=False,
                                                  api_key=self.api_key,
                                                  auth_id=self.xdr_auth_id)
-        custom_user_agent = self.get_custom_user_agent()
+        custom_user_agent = generate_user_agent(self.build_number)
```
Comment (kobymeir) [inline Tests/configure_and_test_integration_instances.py:130]:
```suggestion
        custom_user_agent = get_custom_user_agent(self.build_number)
```

File: Tests/configure_and_test_integration_instances.py [hunk 18]
```diff
@@ -158,7 +158,7 @@ def reconnect_client(self):
                                                  verify_ssl=False,
                                                  username=self.user_name,
                                                  password=self.password)
-        custom_user_agent = self.get_custom_user_agent()
+        custom_user_agent = generate_user_agent(self.build_number)
```
Comment (kobymeir) [inline Tests/configure_and_test_integration_instances.py:161]:
```suggestion
        custom_user_agent = get_custom_user_agent(self.build_number)
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 19]
```diff
@@ -1,15 +1,21 @@
...
+from datetime import datetime, timedelta
+from time import sleep
 
 import demisto_client
-from Tests.configure_and_test_integration_instances import CloudBuild
+from demisto_client.demisto_api.rest import ApiException
+from urllib3.exceptions import HTTPWarning, HTTPError
+
+from Tests.Marketplace.configure_and_install_packs import search_and_install_packs_and_their_dependencies
+from Tests.configure_and_test_integration_instances import CloudBuild, generate_user_agent
```
Comment (kobymeir) [inline Tests/Marketplace/search_and_uninstall_pack.py:14]:
```suggestion
from Tests.configure_and_test_integration_instances import CloudBuild, get_custom_user_agent
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 20]
```diff
@@ -259,6 +379,9 @@ def main():
                                       verify_ssl=False,
                                       api_key=api_key,
                                       auth_id=xdr_auth_id)
+    client.api_client.user_agent = generate_user_agent(options.build_number)
```
Comment (kobymeir) [inline Tests/Marketplace/search_and_uninstall_pack.py:382]:
```suggestion
    client.api_client.user_agent = get_custom_user_agent(options.build_number)
```

File: Tests/Marketplace/search_and_uninstall_pack.py [hunk 21]
```diff
@@ -77,23 +83,140 @@ def uninstall_all_packs_one_by_one(client: demisto_client, hostname, unremovable
...
+            except ApiException as ex:
+                if not attempt:  # exhausted all attempts, understand what happened and exit.
+                    # Unknown exception reason, re-raise.
+                    raise Exception(f"Got status {ex.status} from server, message: {ex.body}, headers: {ex.headers}") from ex
+            except (HTTPError, HTTPWarning) as http_ex:
+                if not attempt:
+                    raise Exception("Failed to perform http request to the server") from http_ex
+
+            # There are more attempts available, sleep and retry.
+            logging.debug(f"failed to get installation/update status, sleeping for {sleep_interval} seconds.")
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_uninstall_pack.py:122]:
```suggestion
            logging.debug(f"Failed to get installation/update status, sleeping for {sleep_interval} seconds.")
```

**Review by tkatzir — APPROVED** [review]
Great! Approved.

One minor comment, not a biggy. Only if you're doing other changes.

**Comment by kobymeir** [comment]
Force merge - testing uninstall packs one by one during the nightly.
Everything passed, except being synced with master.

**Pull request merged**